Using file storage for config storage tree causes some problems

**Describe the bug**

If the config storage tree is configured to use 'file' provider, several errors occur in the app. such as "not a directory path: acls" or "not a directory path: projects/myproject/acls".


~~~
org.rundeck.storage.api.StorageException: not a directory path: acls
	at org.rundeck.storage.api.StorageException.listException(StorageException.java:85)
	at org.rundeck.storage.data.file.FileTree.filterResources(FileTree.java:166)
	at org.rundeck.storage.data.file.FileTree.listDirectory(FileTree.java:120)
	at org.rundeck.storage.impl.DelegateTree.listDirectory(DelegateTree.java:71)
	at org.rundeck.storage.impl.DelegateTree.listDirectory(DelegateTree.java:71)
	at org.rundeck.storage.conf.ListenerTree.listDirectory(ListenerTree.java:97)
	at org.rundeck.storage.impl.DelegateTree.listDirectory(DelegateTree.java:71)
	at org.rundeck.storage.impl.DelegateTree.listDirectory(DelegateTree.java:71)
	at org.rundeck.storage.impl.StringToPathTree.listDirectory(StringToPathTree.java:66)
	at rundeck.services.ConfigStorageService.$tt__listDirPaths(ConfigStorageService.groovy:71)
~~~

Additionally, saving a new project config file fails with "NullPointerException" "Cannot invoke method propertyNames() on null object"

~~~
Caused by: java.lang.NullPointerException: Cannot invoke method propertyNames() on null object
	at rundeck.services.ProjectManagerService.mergeProperties(ProjectManagerService.groovy:603)
	at rundeck.services.ProjectManagerService.$tt__mergeProjectProperties(ProjectManagerService.groovy:588)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at rundeck.services.ProjectManagerService.$tt__mergeProjectProperties(ProjectManagerService.groovy:565)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at com.dtolabs.rundeck.server.projects.RundeckProject.mergeProjectProperties(RundeckProject.java:82)
	at rundeck.services.FrameworkService.updateFrameworkProjectConfig(FrameworkService.groovy:249)
	at rundeck.controllers.FrameworkController.saveProject(FrameworkController.groovy:1215)
	... 83 common frames omitted

~~~

**My Rundeck detail**

* Rundeck version: 3.0

**To Reproduce**

Setup config storage tree using file provider in rundeck-config:

~~~
rundeck.config.storage.provider.1.type=file
rundeck.config.storage.provider.1.path=/
rundeck.config.storage.provider.1.config.baseDir=/tmp/rd-store-configs
~~~

**Expected behavior**

Rundeck works as before, (excepting missing acls/project configs)
**Additional context**

Creating some missing dirs allows certain pages to load, but is not a very workable workaround.

--- a/rundeckapp/grails-app/services/rundeck/services/ConfigStorageService.groovy
+++ b/rundeckapp/grails-app/services/rundeck/services/ConfigStorageService.groovy
@@ -67,8 +67,8 @@ class ConfigStorageService implements StorageManager {
      * @return
      */
     List<String> listDirPaths(String path, String pattern = null) {
-        def storagePath = (path.startsWith("/") ? path : "/${path}")
-        def resources = getStorage().listDirectory(storagePath)
+        def storagePath = (path.startsWith("/") ? path : "/${path}").toString()
+        Collection<Resource<ResourceMeta>> resources = getStorage().hasDirectory(storagePath)?getStorage().listDirectory(storagePath):[]
         def outprefix = path.endsWith('/') ? path.substring(0, path.length() - 1) : path
         resources.collect { Resource<ResourceMeta> res ->
             def pathName = res.path.name + (res.isDirectory() ? '/' : '')

--- a/rundeckapp/grails-app/services/rundeck/services/ProjectManagerService.groovy
+++ b/rundeckapp/grails-app/services/rundeck/services/ProjectManagerService.groovy
@@ -306,7 +306,7 @@ class ProjectManagerService implements ProjectManager, ApplicationContextAware, 
     List<String> listProjectDirPaths(String projectName, String path, String pattern=null) {
         def prefix = 'projects/' + projectName
         def storagePath = prefix + (path.startsWith("/")?path:"/${path}")
-        def resources = getStorage().listDirectory(storagePath)
+        def resources = getStorage().hasDirectory(storagePath)?getStorage().listDirectory(storagePath):[]
         def outprefix=path.endsWith('/')?path.substring(0,path.length()-1):path
         resources.collect{Resource<ResourceMeta> res->
             def pathName=res.path.name + (res.isDirectory()?'/':'')

--- a/rundeckapp/grails-app/services/rundeck/services/ProjectManagerService.groovy
+++ b/rundeckapp/grails-app/services/rundeck/services/ProjectManagerService.groovy
@@ -583,7 +583,7 @@ class ProjectManagerService implements ProjectManager, ApplicationContextAware, 
             throw new IllegalArgumentException("project does not exist: " + projectName)
         }
         def res = loadProjectConfigResource(projectName)
-        def oldprops = res.config
+        Properties oldprops = res?.config?:new Properties()
         Properties newprops = mergeProperties(removePrefixes, oldprops, properties)
         Map newres=storeProjectConfig(projectName, newprops)
         newres

--- a/rundeckapp/src/test/groovy/rundeck/services/ConfigStorageServiceSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/services/ConfigStorageServiceSpec.groovy
@@ -83,6 +83,19 @@ class ConfigStorageServiceSpec extends Specification {
         service.listDirPaths("my-dir")==['my-dir/file1','my-dir/file2','my-dir/etc/']
         service.listDirPaths("not-my-resource")==[]
     }
+
+    void "list dir paths should check dir exists"() {
+        given: 'storage backend which throws exception on listDirectory for non-existent dir'
+        service.rundeckConfigStorageTree = Stub(StorageTree) {
+            hasDirectory('/my-dir') >> false
+            listDirectory('/my-dir') >> {
+                throw StorageException.listException(PathUtil.asPath('/my-dir'), "dne")
+            }
+        }
+        expect: 'result is empty instead of exception'
+        [] == service.listDirPaths('my-dir')
+    }
+    
     void "storage list paths regex"(){
         given:
         service.rundeckConfigStorageTree=Stub(StorageTree){

--- a/rundeckapp/src/test/groovy/rundeck/services/ProjectManagerServiceSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/services/ProjectManagerServiceSpec.groovy
@@ -34,6 +34,7 @@ import org.rundeck.storage.api.Resource
 import org.rundeck.storage.api.StorageException
 import rundeck.Project
 import spock.lang.Specification
+import spock.lang.Unroll
 
 /**
  * See the API for {@link grails.test.mixin.services.ServiceUnitTestMixin} for usage instructions
@@ -445,11 +446,62 @@ class ProjectManagerServiceSpec extends Specification {
         1*service.projectCache.invalidate('test1')
         1*service.nodeService.expireProjectNodes('test1')
 
-        res!=null
-        res.config.size()==3
-        'def'==res.config['abc']
-        'ghi'==res.config['def']
-        'test1'==res.config['project.name']
+        res != null
+        res.config.size() == 3
+        'def' == res.config['abc']
+        'ghi' == res.config['def']
+        'test1' == res.config['project.name']
+    }
+
+    @Unroll
+    void "merge project properties without existing resource"() {
+        setup:
+        Properties props1 = new Properties()
+        props1['def'] = 'ghi'
+        new Project(name: 'test1').save()
+        service.storage = Stub(StorageTree) {
+            hasResource("projects/test1/etc/project.properties") >> false
+            updateResource(
+                "projects/test1/etc/project.properties", { ResourceMeta rm ->
+                def tprops = new Properties()
+                tprops.load(rm.inputStream)
+                rm.meta.size() == 1 &&
+                rm.meta[StorageUtil.RES_META_RUNDECK_CONTENT_TYPE] ==
+                ProjectManagerService.MIME_TYPE_PROJECT_PROPERTIES &&
+                tprops['def'] ==
+                'ghi'
+            }
+            ) >> Stub(Resource) {
+                getContents() >> Stub(ResourceMeta) {
+                    getInputStream() >> new ByteArrayInputStream(
+                        (
+                            '#' + ProjectManagerService.MIME_TYPE_PROJECT_PROPERTIES +
+                            '\ndef=ghi'
+                        ).bytes
+                    )
+                }
+            }
+        }
+
+        service.nodeService = Mock(NodeService)
+        service.projectCache = Mock(LoadingCache)
+
+        when:
+        def res = service.mergeProjectProperties('test1', props1, removePrefixes as Set)
+
+        then:
+        1 * service.projectCache.invalidate('test1')
+        1 * service.nodeService.expireProjectNodes('test1')
+
+        res != null
+        res.config.size() == 2
+        'ghi' == res.config['def']
+        'test1' == res.config['project.name']
+
+        where:
+        removePrefixes       | _
+        ['resources.source'] | _
+        []                   | _
     }
     void "set project properties internal"(){
         setup:

--- a/rundeckapp/src/test/groovy/rundeck/services/ProjectManagerServiceSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/services/ProjectManagerServiceSpec.groovy
@@ -731,6 +731,18 @@ class ProjectManagerServiceSpec extends Specification {
         service.listProjectDirPaths("test1","my-dir")==['my-dir/file1','my-dir/file2','my-dir/etc/']
         service.listProjectDirPaths("test1","not-my-resource")==[]
     }
+
+    void "list project dir paths when tree throws exception"() {
+        given:
+        service.storage = Stub(StorageTree) {
+            hasDirectory("projects/test1/my-dir") >> false
+            listDirectory("projects/test1/my-dir") >> {
+                throw StorageException.listException(PathUtil.asPath('projects/test1/my-dir'), 'dne')
+            }
+        }
+        expect:
+        service.listProjectDirPaths("test1", "my-dir") == []
+    }
     void "storage list paths regex"(){
         given:
         service.storage=Stub(StorageTree){

